dev/src/document.ts

```diff
@@ -23,9 +23,90 @@ import is from 'is';
 import {FieldPath} from './path';
 import {FieldTransform} from './field-value';
 import {Timestamp} from './timestamp';
-import {validatePkg} from './validate';
 import {isPlainObject} from './serializer';
 
+/**
+ * Returns a builder for DocumentSnapshot and QueryDocumentSnapshot instances.
+ * Invoke `.build()' to assemble the final snapshot.
+ *
+ * @private
+ * @class
+ */
+class DocumentSnapshotBuilder {
+  ref;
+  fieldsProto;
+  readTime;
+  createTime;
+  updateTime;
+  /**
+   * @private
+   * @hideconstructor
+   *
+   * @param {DocumentSnapshot=} snapshot An optional snapshot to base this
+   * builder on.
+   */
+  constructor(snapshot?) {
```

Comment:
`snapshot?:DocumentSnapshot`

Reply:
I'm not looking to get to complete typedness here (FWIW, this change is complex because `DocumentSnapshot` has a bunch of required fields and methods atm)